File: backend/src/main/kotlin/com/casadosreclamos/dto/RequestSlotDto.kt
package com.casadosreclamos.dto

import com.casadosreclamos.model.request.Measurements
import com.casadosreclamos.model.request.RequestSlot
import java.util.stream.Collectors

data class RequestSlotDto(
    var image: ImageDto?,
    var measurements: Measurements?,
    var material: MaterialDto?,
    var finishings: MutableSet<FinishingDto>?
) {
    constructor(slot: RequestSlot) : this(
        ImageDto(slot.image),
        slot.measurements,
        MaterialDto(slot.material),
        slot.finishings.stream().map { FinishingDto(it) }.collect(
            Collectors.toSet()
        )
    )
}

File: frontend/src/models/dto/RequestSlotDto.ts
import FinishingDto from "@models/dto/FinishingDto";
import ImageDto from "@models/dto/ImageDto";
import MaterialDto from "@models/dto/MaterialDto";
import Measurements from "@models/dto/Measurements";

export default class RequestSlotDto {
  image: ImageDto;
  measurements: Measurements;
  material: MaterialDto;
  finishings: Array<FinishingDto>;

  constructor(obj: RequestSlotDto) {
    this.image = new ImageDto(obj.image);
    this.measurements = new Measurements(obj.measurements);
    this.material = new MaterialDto(obj.material);
    this.finishings = obj.finishings.map((f) => new FinishingDto(f));
  }

  getMaterial(): string {
    return this.material.name;
  }

  getFinishings(): string[] {
    return this.finishings.map((f) => f.name);
  }

  getMeasurements(): number[] {
    return [this.measurements.width, this.measurements.height];
  }
}
